File: Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/BaseRepositoryImpl.java
package com.habitrpg.android.habitica.data.implementation;

import com.habitrpg.android.habitica.data.ApiClient;
import com.habitrpg.android.habitica.data.BaseRepository;
import com.habitrpg.android.habitica.data.local.BaseLocalRepository;

import java.util.List;

import io.realm.RealmObject;

public abstract class BaseRepositoryImpl<T extends BaseLocalRepository> implements BaseRepository {

    protected final T localRepository;
    protected final ApiClient apiClient;

    public BaseRepositoryImpl(T localRepository, ApiClient apiClient) {
        this.localRepository = localRepository;
        this.apiClient = apiClient;
    }

    public void close() {
        this.localRepository.close();
    }

    @Override
    public <T extends RealmObject> List<T> getUnmanagedCopy(List<T> list) {
        return localRepository.getUnmanagedCopy(list);
    }

    @Override
    public <T extends RealmObject> T getUnmanagedCopy(T object) {
        return localRepository.getUnmanagedCopy(object);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/data/implementation/BaseRepositoryImpl.kt
package com.habitrpg.android.habitica.data.implementation

import com.habitrpg.android.habitica.data.ApiClient
import com.habitrpg.android.habitica.data.BaseRepository
import com.habitrpg.android.habitica.data.local.BaseLocalRepository
import com.habitrpg.android.habitica.models.BaseObject
import com.habitrpg.android.habitica.modules.AuthenticationHandler

abstract class BaseRepositoryImpl<T : BaseLocalRepository>(
    protected val localRepository: T,
    protected val apiClient: ApiClient,
    protected val authenticationHandler: AuthenticationHandler
) : BaseRepository {

    val currentUserID: String
        get() = authenticationHandler.currentUserID ?: ""

    override fun close() {
        this.localRepository.close()
    }

    override fun <T : BaseObject> getUnmanagedCopy(list: List<T>): List<T> {
        return localRepository.getUnmanagedCopy(list)
    }

    override val isClosed: Boolean
        get() = localRepository.isClosed

    override fun <T : BaseObject> getUnmanagedCopy(obj: T): T {
        return localRepository.getUnmanagedCopy(obj)
    }
}
